[iOS] [SaferC++] Improve memory safety in WebPage/ and WebCoreSupport/

zakariaridouh opened the pull request:
#### 9a8d9f0b2cb409f95cf41fcc95e14c83325a551f
<pre>
[iOS] [SaferC++] Improve memory safety in WebPage/ and WebCoreSupport/
<a href="https://bugs.webkit.org/show_bug.cgi?id=301642">https://bugs.webkit.org/show_bug.cgi?id=301642</a>
<a href="https://rdar.apple.com/163656764">rdar://163656764</a>

Reviewed by Anne van Kesteren.

Fix failing files in WebPage/ and WebCoreSupport/

* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKit/WebProcess/WebCoreSupport/WebPlatformStrategies.cpp:
(WebKit::WebPlatformStrategies::writeToPasteboard):
(WebKit::WebPlatformStrategies::updateSupportedTypeIdentifiers):
* Source/WebKit/WebProcess/WebPage/ViewGestureGeometryCollector.cpp:
(WebKit::ViewGestureGeometryCollector::computeTextLegibilityScales):
(WebKit::ViewGestureGeometryCollector::computeMinimumAndMaximumViewportScales const):

Canonical link: <a href="https://commits.webkit.org/302335@main">https://commits.webkit.org/302335@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2e0efc652b37591ee2c8a62f160969a6c4683125

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128710 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/966 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39539 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136093 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80097 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/09b0eb16-d92f-4eb3-92f5-f2d737fdd2c8) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/919 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/846 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97984 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65895 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/4901cfa4-fc1e-43ad-8b0d-1efac341ef27) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131658 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/687 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115305 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78599 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/dde95034-6657-4559-984a-61908fa84078) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/638 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79372 "Built successfully") | | 
| | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109065 "Build is in progress. Recent messages:OS: Tahoe (26.0), Xcode: 26.0.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; 8 api tests failed or timed out; Running re-run-api-tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33900 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138550 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/789 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/795 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106518 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/841 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111644 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106338 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27090 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/660 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30175 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/53171 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/854 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64120 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/713 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/769 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/800 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebKit/WebProcess/WebPage/ViewGestureGeometryCollector.cpp:
```diff
@@ -198,12 +198,12 @@ std::optional<std::pair<double, double>> ViewGestureGeometryCollector::computeTe
         if (!is<Text>(documentTextIterator.node()))
             continue;
 
-        auto& textNode = downcast<Text>(*documentTextIterator.node());
-        auto textLength = textNode.length();
-        if (!textLength || !textNode.renderer() || allTextNodes.contains(textNode))
+        Ref textNode = downcast<Text>(*documentTextIterator.node());
```
Kinda surprised we use `downcast` and not `uncheckedDowncast` here.

zakariaridouh replied:
Not sure! Thank you for the review :)

cdumez replied:
Ideally we'd use dynamicDowncast above instead of `is<>`. This is inefficient.

annevk replied:
I was thinking of suggesting that but then I worried about the cost of the variable. I'll PR.

cdumez replied:
Maybe you saw something I didn't? I thought we already had a local `textNode` variable here anyway.

annevk replied:
Does it not make a difference that it comes after the `continue` currently?

annevk replied:
https://github.com/WebKit/WebKit/pull/53192

zakariaridouh committed:
[iOS] [SaferC++] Improve memory safety in WebPage/ and WebCoreSupport/
https://bugs.webkit.org/show_bug.cgi?id=301642
rdar://163656764

Reviewed by Anne van Kesteren.

Fix failing files in WebPage/ and WebCoreSupport/

* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKit/WebProcess/WebCoreSupport/WebPlatformStrategies.cpp:
(WebKit::WebPlatformStrategies::writeToPasteboard):
(WebKit::WebPlatformStrategies::updateSupportedTypeIdentifiers):
* Source/WebKit/WebProcess/WebPage/ViewGestureGeometryCollector.cpp:
(WebKit::ViewGestureGeometryCollector::computeTextLegibilityScales):
(WebKit::ViewGestureGeometryCollector::computeMinimumAndMaximumViewportScales const):

Canonical link: https://commits.webkit.org/302335@main
